File: c++/2024.cpp
// Link: https://leetcode.com/problems/maximize-the-confusion-of-an-exam/description/
// Runtime: 76 ms
// Memory: 9 MB

class Solution {
public:
  int maxConsecutiveAnswers(const std::string &answerKey, int k) {
    int maxSize = 0;
    std::unordered_map<char, int> occurances;
    for (int right = 0; right < static_cast<int>(answerKey.size()); right++) {
      occurances[answerKey[right]]++;
      if (std::min(occurances['T'], occurances['F']) <= k)
        maxSize++;
      else
        occurances[answerKey[right - maxSize]]--;
    }

    return maxSize;
  }
};

File: python/2024.py
# Link: https://leetcode.com/problems/maximize-the-confusion-of-an-exam/description/
# Runtime: 278 ms
# Memory: 16.8 MB


class Solution:
    def maxConsecutiveAnswers(self, answerKey: str, k: int) -> int:
        maxSize = 0
        occurances = defaultdict(lambda: 0)
        for right, value in enumerate(answerKey):
            occurances[value] += 1
            if min(occurances["T"], occurances["F"]) <= k:
                maxSize += 1
            else:
                occurances[answerKey[right - maxSize]] -= 1

        return maxSize
